File: 1819-construct-the-lexicographically-largest-valid-sequence/construct-the-lexicographically-largest-valid-sequence.py
class Solution:
    def constructDistancedSequence(self, n: int) -> List[int]:
        seq = [0] * (2 * n - 1)
        seen = set()
        def dfs(i):
            if i == len(seq):
                return True
            if seq[i]:
                return dfs(i + 1)
            for j in range(n, 0, -1):
                if j in seen:
                    continue
                seen.add(j)
                seq[i] = j
                if j == 1:
                    if dfs(i + 1):
                        return True
                elif i + j < len(seq) and seq[i + j] == 0:
                    seq[i + j] = j
                    if dfs(i + 1):
                        return True
                    seq[i + j] = 0
                seq[i] = 0
                seen.remove(j)
            return False
        dfs(0)
        return seq

File: 1819-construct-the-lexicographically-largest-valid-sequence/construct-the-lexicographically-largest-valid-sequence.java
class Solution {
    public int[] constructDistancedSequence(int n) {
        int len = 2 * n - 1;
        int[] seq = new int[len];

        boolean[] used = new boolean[n + 1];

        dfs(0, seq, used, n);
        return seq;
    }

    private boolean dfs(int index, int[] seq, boolean[] used, int n) {
        if (index == seq.length) {
            return true;
        }

        if (seq[index] != 0) {
            return dfs(index + 1, seq, used, n);
        }

        for (int j  = n; j >= 1; j--) {
            if (used[j]) {
                continue;
            }

            used[j] = true;
            seq[index] = j;

            if (j == 1) {
                if (dfs(index + 1, seq, used, n)) {
                    return true;
                }
            } else {
                if (index + j < seq.length && seq[index + j] == 0) {
                    seq[index + j] = j;
                    if (dfs(index + 1, seq, used, n)) {
                        return true;
                    }

                    seq[index + j] = 0;
                }
            }

            seq[index] = 0;
            used[j] = false;
        }
        return false;
    }
}